### ISSUE TITLE
Brightness range is incorrect, doesn't turn off on zero

### ISSUE THREAD

**Issue by hoffbaked** [tuya/tuya-home-assistant#315]
**Describe the bug**
All the remap calls in light.py for brightness have 0 to 255 as the range, when 1 is the minimum for Home Assistant. Setting a brightness value of 1 should map to the minimum.

**Expected behavior**
Setting a brightness value of zero should turn off the light. 1 should set the minimum value. Additionally, if the Tuya light is at its minimum value, its brightness should show as 1 in Home Assistant.

**Home Assistant Version**
- 2021.8.0

**Comment by tsutsuku** [comment]
According to LightEntity, valid brightness is 0~255.

<img width="802" alt="WeChatWorkScreenshot_cc530e78-c6fa-4356-a097-5c9a3bef487c" src="https://user-images.githubusercontent.com/18442047/128443777-37fb67a6-07a2-42fd-b8c4-f7de88bb3db0.png">

**Comment by hoffbaked** [comment]
Yes, but a value of 0 turns off the light, and a value of 1 is the minimum brightness. 

—David

**Comment by hoffbaked** [comment]
Here’s the description for the parameters for the light.turn_on service in the Home Assistant UI:



—David

**Comment by hoffbaked** [comment]
Oops. I attatched something to an email reply, and it didn’t show up. Here’s the description of the brightness parameter from Home Assistant on the parameter for the light.turn_on service:
![Home Assistant Brightness Description](https://user-images.githubusercontent.com/17089581/134602034-f45cac70-19bb-4d11-8a69-7ece23e410ad.png)

**Comment by pabigot** [comment]
I'm not convinced this is a bug in tuya_v2.  It may be an error in the documentation of the turn_on service.  Or it may be accurate, but that service does something special and invokes `Light.turn_off` rather than `Light.turn_on` if `brightness` is zero.

The [light entity](https://developers.home-assistant.io/docs/core/entity/light/) documentation for the brightness property doesn't clearly state that a zero turns the light off.  That is a reasonable expectation, but so is "zero means turn the bulb on at its lowest supported intensity" (for [Tuya](https://developer.tuya.com/en/docs/iot/f?id=K9i5ql3v98hn3) that's either `bright_value=25` (/255) or `bright_value_v2=10` (/1000).

Nor does the light entity documentation for the `turn_on` method clearly specify that turning on with zero is the same as turning off.

I only have Tuya and Tasmota bulbs available: for both, invoking a scene with these properties:

```
    state: 'on'
    color_mode: hs
    hs_color: [60, 100]
    brightness: 0
```

turns the bulb green on at the lowest supported brightness.

**Comment by hoffbaked** [comment]
I tried calling light.turn_on with a brightness value of zero, and for ZHA and zwave_js devices, they turn off, as per the descriptions in all the documentation. 

—David

**Comment by hoffbaked** [comment]
It is clearly stated on this page:
https://www.home-assistant.io/integrations/light/

And I just check some devices I have on the old deprecated Z-Wave integration, and they also turn off with a zero brightness.

**Comment by pabigot** [comment]
> I tried calling light.turn_on with a brightness value of zero, and for ZHA and zwave_js devices, they turn off, as per the descriptions in all the documentation. —David

But is that because the devices in question are actually turned off (the interior state of the device is off), or because they accept zero as brightness (and do not emit any light) while the interior state is actually on?

In other words, is the behavior difference due to different hardware capabilities, or because software layers are intentionally converting a command to turn the device on into a command to turn the device off?

Or perhaps Z-Wave lights require "zero brightness means off" by standard, while Wi-Fi ones do not.

> It is clearly stated on this page:
> https://www.home-assistant.io/integrations/light/

But https://www.home-assistant.io/integrations/light/#service-lightturn_on documents  the expected behavior of the `light.turn_on` **service** (which supports operating on potentially heterogeneous groups of abstract entities), not the required behavior for [method](https://developers.home-assistant.io/docs/core/entity/light/#turn-on-light-device) `turn_on` in a specific integration's implementation of the `Light` **entity**.

I can see a higher-level service having a responsibility to provide consistency across integrations, while direct commands to a specific integration might provide more flexibility to support vendor features.  The following requirements are both independently justifiable:

* Entity method `Light.turn_on` with `brightness=0` shall be functionally equivalent to entity method `Light.turn_off`
* Entity method `Light.turn_on` with `brightness=0` shall turn the device on with the lowest supported brightness level

The second allows gemeric description of a desired state that can't be provided by the first.

In any case, I've used Home Assistant for less than a week and am still learning.  You're probably right about how things are supposed to be done.  I'm commenting only to document why, after reading [this comment](https://github.com/tuya/tuya-home-assistant/issues/316#issuecomment-926269641), I won't try to fix this issue in #486.

**Comment by hoffbaked** [comment]
I agree that it’s a bit weird to have brightness 0 be equivalent to turn_off, but the state does indeed turn to off on the three other integrations I have access to. And while the method documentation is unclear, the current implementation causes the service call to behave incorrectly. But I completely understand not feeling comfortable fixing it, especially when your only other integration for testing behaves the same way.

**Comment by pabigot** [comment]
> the state does indeed turn to off on the three other integrations I have access to.

But all three integrations are Z-Wave.

Looking at [zwave-js](https://github.com/home-assistant/core/blob/dev/homeassistant/components/zwave_js/light.py#L268-L270) it appears the way zwave-js lights are turned off is to set brightness to zero.

[ZHA](https://github.com/home-assistant/core/blob/dev/homeassistant/components/zha/light.py#L321-L341) does something similar, and retains the configured brightness setting while turned off, restoring it when turned on, suggesting it is working around the limitation that there is not a separate retained boolean attribute to indicate the switch status.

Tuya and [tasmota](https://github.com/home-assistant/core/blob/dev/homeassistant/components/tasmota/light.py#L279-L288) devices control on-off with a separate state attribute.

I don't think what you see is evidence that tuya_v2 is behaving incorrectly; it's just not doing what Z-Wave does.  Which behavior is correct is unclear; it may even be undefined.

**Comment by hoffbaked** [comment]
I guess it’s possible that all the documentation of the light.turn_on service parameter is incorrect, the UI for Home Assistant is incorrect, and the three integrations I’ve tried are incorrect in acting according to the documentation, and that the vague description of the parameter in the method takes precedent and tuya_2 and tasmota are the only ones who are correct. But, until it’s settled, I can’t access the minium brightness, because I can’t use a brightness of zero with an area as a target, because everything in the room will turn off except for the tuya_v2 devices.

**Comment by pabigot** [comment]
I don't know what's correct, I'm just really concerned that a model that makes "on with brightness zero" equal to "off" would results in inconsistent state on any device that has a distinct internal property indicating on/off state.  But yes, people who are responsible for Home Assistant as a whole need to review all this and decide what's right.

In the meantime, why can't you use `state: 'off'` to turn things off, and `state: 'on' ; brightness: 1` to turn things on with minimum (or almost minimum) brightness?  Something about the way the automation works? (I'm just starting to work with that part of HA.)

**Comment by hoffbaked** [comment]
That’s what I’ve always done, and things like the Adaptive Lighting integration only allow 1 as a minimum. But, the 1 value on the lights in my kitchen were significantly brighter than turning it to the minimum with the Tuya app, especially compared to the Zigbee bulbs in the same room. On my local copy of tuya_v2, I just fixed all the ranges remaps to be 1-255, and didn’t really care about the 0 value, and I was much happier.
